pkg/traces/exporter/exporter.go

```diff
@@ -0,0 +1,191 @@
+	// Set all the attributes that we know we can get first
+	attrs := []attribute.KeyValue{
+		semconv.ServiceName(applicationName),
+		semconv.ServiceVersion(version.Version().Version),
+	}
+
+	if archAttr, ok := archAttributeMap[runtime.GOARCH]; ok {
+		attrs = append(attrs, archAttr)
+	}
+
+	t := &TraceExporter{
+		serverProvidedDataStore: serverProvidedDataStore,
+		osqueryClient:           client,
+		attrs:                   attrs,
+		attrLock:                sync.RWMutex{},
+	}
+
+	t.addDeviceIdentifyingAttributes()
+
+	// Set the provider with as many resource attributes as we could get immediately
+	exp, err := newExporter(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new exporter: %w", err)
+	}
+	t.setNewGlobalProvider(exp)
+
+	// In the background, wait for osquery to be ready so that we can fetch more resource
+	// attributes for our traces, then replace the provider with a new one.
+	go func() {
+		t.addAttributesFromOsquery()
+		if exp, err := newExporter(ctx); err == nil {
+			t.setNewGlobalProvider(exp)
+		}
+	}()
+
+	return t, nil
+}
+
+// newExporter returns an exporter that will send traces with OTLP over HTTP.
+func newExporter(ctx context.Context) (sdktrace.SpanExporter, error) {
+	traceClient := otlptracegrpc.NewClient(otlptracegrpc.WithInsecure())
+	exp, err := otlptrace.New(ctx, traceClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not create exporter for traces: %w", err)
+	}
+
+	return exp, nil
+}
+
+// addDeviceIdentifyingAttributes gets device identifiers from the server-provided
+// data and adds them to our resource attributes.
+func (t *TraceExporter) addDeviceIdentifyingAttributes() {
+	t.attrLock.Lock()
+	defer t.attrLock.Unlock()
+
+	if deviceId, err := t.serverProvidedDataStore.Get([]byte("device_id")); err == nil {
+		t.attrs = append(t.attrs, semconv.ServiceInstanceID(string(deviceId)))
+	}
+
+	if munemo, err := t.serverProvidedDataStore.Get([]byte("munemo")); err == nil {
+		t.attrs = append(t.attrs, attribute.String("launcher.munemo", string(munemo)))
+	}
+
+	if orgId, err := t.serverProvidedDataStore.Get([]byte("organization_id")); err == nil {
+		t.attrs = append(t.attrs, attribute.String("launcher.organization_id", string(orgId)))
+	}
+
+	if serialNumber, err := t.serverProvidedDataStore.Get([]byte("serial_number")); err == nil {
+		t.attrs = append(t.attrs, attribute.String("launcher.serial", string(serialNumber)))
+	}
+}
+
+// addAttributesFromOsquery retrieves device and OS details from osquery and adds them
+// to our resource attributes. Since this is called on startup when the osquery client
+// may not be ready yet, we perform a few retries.
+func (t *TraceExporter) addAttributesFromOsquery() {
+	t.attrLock.Lock()
+	defer t.attrLock.Unlock()
+
+	// The osqueryd client may not have initialized yet, so retry a few times on error.
+	osquerydRetries := 5
+	for i := 0; i < osquerydRetries; i += 1 {
+		query := `
```

Comment:
readability nit, might be easier to understand this code if we put the query as a const outside the for loop. Consider using backoff.watifor. Not that there is anything wrong with the loop, it's just kind of a common practice in our code base.

Reply:
Updated, let me know how it looks to you!

Reply:
look good!